src/compiler.rkt

```diff
@@ -276,356 +276,289 @@
-        (vector-set! vprecs-new (- x varc)
-                     (max ; check whether this op already has a precision that is higher
-                      (vector-ref vprecs-new (- x varc))
-                      child-exponents))))))
-
-; Function calculates an exponent for a certain output and input using condition formulas,
+    ;(define child-exponents (+ exps-from-above new-exponents))
+    (for ([x (in-list tail-registers)]
+          [new-exp (in-list new-exponents)])
+    (when (>= x varc)   ; when tail register is not a variable
+      (vector-set! vprecs-new (- x varc)
+                   (max ; check whether this op already has a precision that is higher
+                    (vector-ref vprecs-new (- x varc))
+                    (+ exps-from-above new-exp)))))))
+
+(define (ival-max-log2-approx x)
+  (max (log2-approx (ival-hi x)) (log2-approx (ival-lo x))))
+(define (ival-min-log2-approx x)
+  (min (log2-approx (ival-hi x)) (log2-approx (ival-lo x))))
+
+; Function calculates an exponent per input for a certain output and inputs using condition formulas,
 ;   where an exponent is an additional precision that needs to be added to srcs evaluation so,
 ;   that the output will be fixed in its precision when evaluating again
 (define (get-exponent op output srcs)
-  (cond
-    [(member op (list ival-mult ival-div ival-sqrt ival-cbrt))
-     1]
-    [(equal? op ival-add)
-     ; log[Г+] = max(log[x], log[y]) - log[x + y]
-     ;                               ^^^^^^^^^^^^
-     ;                               possible cancellation
+  (case (object-name  op)
+    [(ival-mult ival-div ival-sqrt ival-cbrt)
+     ; log[Г*]'x = log[Г*]'y = log[Г/]'x = log[Г/]'y = 1
+     ; log[Гsqrt] = 0.5
+     ; log[Гcbrt] = 0.3
+     (make-list (length srcs) 1)]
+    
+    [(ival-add ival-sub)
+     ; log[Г+]'x = log[x] - log[x + y] + 1 (1 for mantissa approximation when dividing)
+     ; log[Г+]'y = log[y] - log[x + y] + 1
+     ; log[Г-]'x = log[x] - log[x - y] + 1
+     ; log[Г-]'y = log[y] - log[x - y] + 1
+     ;                    ^^^^^^^^^^^^
+     ;                    possible cancellation
      (define x (first srcs))
-     (define xlo (ival-lo x))
-     (define xlo-exp (log2-approx xlo))
-     (define xlo-sgn (bigfloat-signbit xlo))
-     (define xhi (ival-hi x))
-     (define xhi-exp (log2-approx xhi))
-     (define xhi-sgn (bigfloat-signbit xhi))
-
      (define y (second srcs))
-     (define ylo (ival-lo y))
-     (define ylo-exp (log2-approx ylo))
-     (define ylo-sgn (bigfloat-signbit ylo))
-     (define yhi (ival-hi y))
-     (define yhi-exp (log2-approx yhi))
-     (define yhi-sgn (bigfloat-signbit yhi))
-     
      (define outlo (ival-lo output))
-     (define outlo-exp (log2-approx outlo))
      (define outhi (ival-hi output))
-     (define outhi-exp (log2-approx outhi))
+     
+     (define x-exp (ival-max-log2-approx x))
+     (define y-exp (ival-max-log2-approx y))
+     (define out-exp (min (log2-approx outhi) (log2-approx outlo)))
```

Comment:
Is this just `(ival-min-log2-approx output)`

Reply:
Yeah, I just evaluated `outhi` and `outlo` above, I didn't want to do additional allocation

Reply:
Well, it's not an allocation, it's a pointer access. Eh. Either way is fine, I'd go with the simpler option, `backward-pass` is not ultimately that expensive anyway.